### ISSUE TITLE
Error in devel with latest IPv8

### ISSUE THREAD

**Issue by qstokkink** [Tribler/tribler#4227]
```
Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

Unhandled error in Deferred:

Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table

ERROR   1550311612.48            tribler_window:73   (root)  Traceback (most recent call last):
  File "/home/quinten/Documents/tribler/TriblerGUI/event_request_manager.py", line 133, in on_read_data
    raise RuntimeError(json_dict["event"]["text"])
RuntimeError: [Failure instance: Traceback (failure with no frames): <type 'exceptions.RuntimeError'>: No nodes found in the routing table
]
```

**Comment by qstokkink** [comment]
Reproduced this with `Deferred.debug` enabled:

```python
Unhandled error in Deferred:
(debug:  C: Deferred was created:
 C:  File "run_tribler.py", line 94, in <module>
 C:    start_tribler_core(base_path, api_port)
 C:  File "run_tribler.py", line 86, in start_tribler_core
 C:    reactor.run()
 C:  File "/home/quinten/.local/lib/python2.7/site-packages/twisted/internet/base.py", line 1267, in run
 C:    self.mainLoop()
 C:  File "/home/quinten/.local/lib/python2.7/site-packages/twisted/internet/base.py", line 1276, in mainLoop
 C:    self.runUntilCurrent()
 C:  File "/home/quinten/.local/lib/python2.7/site-packages/twisted/internet/base.py", line 875, in runUntilCurrent
 C:    f(*a, **kw)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/messaging/interfaces/endpoint.py", line 44, in _deliver_later
 C:    listener.on_packet(packet)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/community.py", line 315, in on_packet
 C:    self.decode_map[msg_id](source_address, data)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 40, in wrapper
 C:    return f(self, *args, **kwargs)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/lazy_community.py", line 137, in wrapper
 C:    return inner_wrapper(self, source_address, data)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/lazy_community.py", line 104, in wrapper
 C:    return func(self, source_address, *unpacked)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/lazy_community.py", line 135, in inner_wrapper
 C:    return func(inner_self, inner_source_address, *combo)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 600, in received_crawl_response
 C:    self.received_half_block(source_address, data[:-12])  # We cut off a few bytes to make it a BlockPayload
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 40, in wrapper
 C:    return f(self, *args, **kwargs)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/lazy_community.py", line 104, in wrapper
 C:    return func(self, source_address, *unpacked)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 295, in received_half_block
 C:    self.process_half_block(block, peer)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 40, in wrapper
 C:    return f(self, *args, **kwargs)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 369, in process_half_block
 C:    validation = self.validate_persist_block(blk)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 345, in validate_persist_block
 C:    self.notify_listeners(block)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 355, in notify_listeners
 C:    listener.received_block(block)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/REST/dht_endpoint.py", line 51, in received_block
 C:    self.publish_latest_block()
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/REST/dht_endpoint.py", line 143, in publish_latest_block
 C:    deferred = self.dht.find_values(self._hashed_dht_key)
 C:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/dht/community.py", line 414, in find_values
 C:    return self._find(target, force_nodes=False)
 C:  File "/home/quinten/.local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1613, in unwindGenerator
 C:    return _cancellableInlineCallbacks(gen)
 C:  File "/home/quinten/.local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1510, in _cancellableInlineCallbacks
 C:    deferred = Deferred(cancel)
 I: First Invoker was:
 I:  File "run_tribler.py", line 94, in <module>
 I:    start_tribler_core(base_path, api_port)
 I:  File "run_tribler.py", line 86, in start_tribler_core
 I:    reactor.run()
 I:  File "/home/quinten/.local/lib/python2.7/site-packages/twisted/internet/base.py", line 1267, in run
 I:    self.mainLoop()
 I:  File "/home/quinten/.local/lib/python2.7/site-packages/twisted/internet/base.py", line 1276, in mainLoop
 I:    self.runUntilCurrent()
 I:  File "/home/quinten/.local/lib/python2.7/site-packages/twisted/internet/base.py", line 875, in runUntilCurrent
 I:    f(*a, **kw)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/messaging/interfaces/endpoint.py", line 44, in _deliver_later
 I:    listener.on_packet(packet)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/community.py", line 315, in on_packet
 I:    self.decode_map[msg_id](source_address, data)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 40, in wrapper
 I:    return f(self, *args, **kwargs)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/lazy_community.py", line 137, in wrapper
 I:    return inner_wrapper(self, source_address, data)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/lazy_community.py", line 104, in wrapper
 I:    return func(self, source_address, *unpacked)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/lazy_community.py", line 135, in inner_wrapper
 I:    return func(inner_self, inner_source_address, *combo)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 600, in received_crawl_response
 I:    self.received_half_block(source_address, data[:-12])  # We cut off a few bytes to make it a BlockPayload
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 40, in wrapper
 I:    return f(self, *args, **kwargs)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/lazy_community.py", line 104, in wrapper
 I:    return func(self, source_address, *unpacked)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 295, in received_half_block
 I:    self.process_half_block(block, peer)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 40, in wrapper
 I:    return f(self, *args, **kwargs)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 369, in process_half_block
 I:    validation = self.validate_persist_block(blk)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 345, in validate_persist_block
 I:    self.notify_listeners(block)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/attestation/trustchain/community.py", line 355, in notify_listeners
 I:    listener.received_block(block)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/REST/dht_endpoint.py", line 51, in received_block
 I:    self.publish_latest_block()
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/REST/dht_endpoint.py", line 143, in publish_latest_block
 I:    deferred = self.dht.find_values(self._hashed_dht_key)
 I:  File "/home/quinten/Documents/tribler/Tribler/pyipv8/ipv8/dht/community.py", line 414, in find_values
 I:    return self._find(target, force_nodes=False)
 I:  File "/home/quinten/.local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1613, in unwindGenerator
 I:    return _cancellableInlineCallbacks(gen)
 I:  File "/home/quinten/.local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1529, in _cancellableInlineCallbacks
 I:    _inlineCallbacks(None, g, status)
 I:  File "/home/quinten/.local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1461, in _inlineCallbacks
 I:    status.deferred.callback(e.value)
)
Traceback (most recent call last):
Failure: exceptions.RuntimeError: No nodes found in the routing table
```

**Comment by qstokkink** [comment]
I assume this should fix it:

```patch
diff --git a/ipv8/REST/dht_endpoint.py b/ipv8/REST/dht_endpoint.py
index 0ed8ffe..3352e81 100644
--- a/ipv8/REST/dht_endpoint.py
+++ b/ipv8/REST/dht_endpoint.py
@@ -15,6 +15,7 @@ from ..attestation.trustchain.listener import BlockListener
 from ..attestation.trustchain.payload import DHTBlockPayload
 from ..messaging.serialization import Serializer
 from ..keyvault.public.libnaclkey import LibNaCLPK
+from ..util import addCallback
 
 
 class DHTEndpoint(resource.Resource):
@@ -141,7 +142,7 @@ class DHTBlockEndpoint(resource.Resource, BlockListener):
                 self.block_version += 1
 
             deferred = self.dht.find_values(self._hashed_dht_key)
-            deferred.addCallback(on_success)
+            addCallback(deferred, on_success)
 
     def render_GET(self, request):
         """
```

**Comment by egbertbouman** [comment]
@qstokkink Introduced by https://github.com/Tribler/py-ipv8/pull/349/files. You're right, it just needs an errback.

**Comment by qstokkink** [comment]
Found another one after running a while, switching `debug = True` again..

**Comment by qstokkink** [comment]
I never managed to reproduce the 2nd failure again, but upon inspection, it might be this line:

https://github.com/Tribler/py-ipv8/blob/49f31b9beb776571b8523ed5ae0ef327014a4b64/ipv8/REST/dht_endpoint.py#L139

@egbertbouman that should also have an errback handler right?

**Comment by egbertbouman** [comment]
@qstokkink Yup, all networking calls to the DHTCommunity need an errback. 

Looking at this code I do wonder about the chances of everything going OK. I see no retry mechanism, and the blocklistener seems to get notified only once about a new block.. Also note that values in the DHT are currently not automatically republished, meaning that the value will dissapear after relatively short period.

Additionally, if the number of chunks grow beyond MAX_VALUES_IN_FIND, you'll never be able to get the full block from the DHT. Then again, blocks are probably small.

**Comment by qstokkink** [comment]
I'll put those improvements in an issue in IPv8 then. For now I just want it to not crash Tribler 😃
